### PR TITLE
C51-382: Folder Structure

### PR DESCRIPTION
**Pull request by deb1990**
## Overview
Civicase previously did not have a proper folder structure for Javascript files. All files were dumped inside the `ang/civicase` folder. This was organisational problem as it was very hard to find a single file when needed.
As part of the PR, a folder structure has been introduced to fix the above mentioned problem.

### Folder Structure
A folder structure similar to CiviHR(https://compucorp.atlassian.net/wiki/spaces/PCHR/pages/121371880/AngularJS) has been followed with some minor exceptions.
1. Each feature does not have its separate module. But this will be fixed in future. As this is not very important, it has been skipped due to project timeline.

### Current Feature Tree
```
├── activity
│   ├── actions
│   ├── calendar
│   ├── card
│   ├── feed
│   ├── filters
│   ├── panel
├── bulk-action
├── case
│   ├── actions
│   ├── card
│   ├── contact-case-tab
│   ├── details
│   │   ├── files-tab
│   │   ├── people-tab
│   │   ├── summary-tab
│   ├── list
│   ├── search
├── contact
├── contact-case-tab
├── dashboard
├── shared
└── app.route.js
```

### Other Changes
1. Now every file contains a single angular js component. That means, files containing multiple components has been broken into individual files.

2. A new Directive has been created for `civicase-dashboard`.

3. Some unused files/code has been removed.
    * `TypeDurationChart.js`
    * `TypeDonutChart.js`
    * `civicaseInteger `
    * ...
4. Unit Test files has been restructured in a similar manner.

5. `ContactsDataService` has been renamed to `ContactsCache`.

6. CiviCRM uses PHP's `glob` function to include multiple files, but `glob` does not support recursive search. To go around this problem, a new function `glob_recursive` is defined, and used inside `ang/civicase.ang.php` to include all JS files.

### CODE REVIEW

File: ang/civicase.ang.php [hunk 1]
```diff
@@ -232,12 +232,34 @@
 $options['allowMultipleCaseClients'] = (bool) Civi::settings()->get('civicaseAllowMultipleClients');
 $options['allowCaseLocks'] = (bool) Civi::settings()->get('civicaseAllowCaseLocks');
 
-return array(
-  'js' => array(
+if ( ! function_exists('glob_recursive')) {
```
Comment (swastikpareek) [inline ang/civicase.ang.php]: Sorry, but seems like we have left a space after `!`. Is it some standard in PHP ?

File: ang/civicase.ang.php [hunk 2]
```diff
@@ -232,12 +232,34 @@
...
+  /**
+   * Recursive Glob function
+   * Source: http://php.net/manual/en/function.glob.php#106595
+   * Does not support flag GLOB_BRACE
+   */
+  function glob_recursive($pattern, $flags = 0) {
+    $files = glob($pattern, $flags);
```
Comment (swastikpareek) [inline ang/civicase.ang.php]: Can we have a new line after variable declarations?

File: ang/civicase.ang.php [hunk 3]
```diff
@@ -232,12 +232,34 @@
...
+    foreach (glob(dirname($pattern).'/*', GLOB_ONLYDIR|GLOB_NOSORT) as $dir) {
+      $files = array_merge($files, glob_recursive($dir.'/'.basename($pattern), $flags));
```
Comment (swastikpareek) [inline ang/civicase.ang.php]: Add a space after and before `.` operator.
```
$files = array_merge($files, glob_recursive($dir . '/' . basename($pattern), $flags));
```

File: ang/civicase.ang.php [hunk 4]
```diff
@@ -232,12 +232,34 @@
...
+ * Get a list of JS files
+ */
+function getJSFiles () {
+  return array_merge(array(
     'assetBuilder://visual-bundle.js', // at the moment, it's safe to include this multiple times -- deduped by resource manager
-    'ang/civicase.js',
-    'ang/civicase/*.js',
-  ),
+    'ang/civicase.js'
+  ), glob_recursive(dirname(__FILE__).'/civicase/*.js'));
```
Comment (swastikpareek) [inline ang/civicase.ang.php]: Same as above

File: ang/civicase.ang.php [hunk 5]
```diff
@@ -232,12 +232,34 @@
...
-  'js' => array(
+if ( ! function_exists('glob_recursive')) {
+  /**
+   * Recursive Glob function
+   * Source: http://php.net/manual/en/function.glob.php#106595
+   * Does not support flag GLOB_BRACE
+   */
+  function glob_recursive($pattern, $flags = 0) {
+    $files = glob($pattern, $flags);
+    foreach (glob(dirname($pattern).'/*', GLOB_ONLYDIR|GLOB_NOSORT) as $dir) {
```
Comment (swastikpareek) [inline ang/civicase.ang.php]: Add a space after and before . operator.

File: ang/civicase/activity/factories/get-activity-feed-url.factory.js [hunk 6]
```diff
@@ -0,0 +1,40 @@
+(function (angular, $, _, CRM) {
+  var module = angular.module('civicase');
+
+  module.factory('getActivityFeedUrl', function ($route, $location, $sce) {
+    return function (caseId, category, statusType, status, id) {
+      caseId = parseInt(caseId, 10);
```
Comment (swastikpareek) [inline ang/civicase/activity/factories/get-activity-feed-url.factory.js]: I know we have moved this code from somewhere, but can we segregate the variables together in a way that new vars are declared/defined before existing vars
```
var af = {};
var currentPath = $location.path();
caseId = parseInt(caseId, 10);
```
And also add a new line after vars declarations and ending if statements.

File: ang/civicase/activity/panel/directives/activity-panel.directive.html [hunk 7]
```diff
@@ -76,7 +76,7 @@ <h3 class="panel-subtitle clearfix">
           <!-- End - Activity Date -->
 
           <!-- Avatar -->
-          <span ng-if="activity.assignee_contact_name" contact-card avatar="true" contacts="activity.assignee_contact_name"></span>
+          <span ng-if="activity.assignee_contact_name" civicase-contact-card avatar="true" contacts="activity.assignee_contact_name"></span>
```
Comment (swastikpareek) [inline ang/civicase/activity/panel/directives/activity-panel.directive.html]: Just a thought. Can we break it down into multiple lines, as it crosses the *80 letters in one line* limit.

File: ang/test/civicase/case/list/directives/case-list-sort-header.directive.spec.js [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    describe('when sortable is false', function () {
+      beforeEach(function () {
+        scope.sort = { field: '', dir: '', sortable: false };
+        header = {
+          display_type: 'activity_card',
+          label: 'Next Activity',
+          name: 'next_activity',
+          sort: 'next_activity'
+        };
+        compileDirective(scope, header);
```
Comment (swastikpareek) [inline ang/test/civicase/case/list/directives/case-list-sort-header.directive.spec.js]: Add new line after var declarations. Check this at other places too

File: ang/civicase.ang.php (hunk 1 above)
Comment (deb1990) [reply ang/civicase.ang.php]: No, its a mistake, fixed. Thanks for noticing.

File: ang/civicase/activity/factories/get-activity-feed-url.factory.js (hunk 6 above)
Comment (deb1990) [reply ang/civicase/activity/factories/get-activity-feed-url.factory.js]: I would rather not fix any exiting code, but as the number of comments are low, I am working on the fixes :)

File: ang/civicase/activity/panel/directives/activity-panel.directive.html (hunk 7 above)
Comment (deb1990) [reply ang/civicase/activity/panel/directives/activity-panel.directive.html]: I would rather not fix any exiting code, but as the number of comments are low, I am working on the fixes :)

File: ang/test/civicase/case/list/directives/case-list-sort-header.directive.spec.js (hunk 8 above)
Comment (deb1990) [reply ang/test/civicase/case/list/directives/case-list-sort-header.directive.spec.js]: I would rather not fix any exiting code, but as the number of comments are low, I am working on the fixes :)

File: ang/civicase.ang.php [hunk 9]
```diff
@@ -232,16 +232,17 @@
...
    * Source: http://php.net/manual/en/function.glob.php#106595
    * Does not support flag GLOB_BRACE
    */
   function glob_recursive($pattern, $flags = 0) {
     $files = glob($pattern, $flags);
-    foreach (glob(dirname($pattern).'/*', GLOB_ONLYDIR|GLOB_NOSORT) as $dir) {
-      $files = array_merge($files, glob_recursive($dir.'/'.basename($pattern), $flags));
+
+    foreach (glob(dirname($pattern) . '/*', GLOB_ONLYDIR|GLOB_NOSORT) as $dir) {
+      $files = array_merge($files, glob_recursive($dir. '/'. basename($pattern), $flags));
```
Comment (swastikpareek) [inline ang/civicase.ang.php]: I am afraid but we still need to add a space before and after `. '/'` 
It should be 
```php
($dir . '/' . basename($pattern), $flags));
```

**Pull request merged**